components/raftstore-v2/src/operation/query/mod.rs

```diff
@@ -71,6 +74,156 @@ impl<'a, EK: KvEngine, ER: RaftEngine, T: raftstore::store::Transport>
         }
     }
 
+    #[inline]
+    pub fn on_refresh_region_buckets(
+        &mut self,
+        region_epoch: RegionEpoch,
+        mut buckets: Vec<raftstore::store::Bucket>,
+        bucket_ranges: Option<Vec<raftstore::store::BucketRange>>,
+    ) {
+        // bucket version layout
+        //   term       logical counter
+        // |-----------|-----------|
+        //  high bits     low bits
+        // term: given 10s election timeout, the 32 bit means 1362 year running time
+        let gen_bucket_version = |term, current_version| {
+            let current_version_term = current_version >> 32;
+            let bucket_version: u64 = if current_version_term == term {
+                current_version + 1
+            } else {
+                if term > u32::MAX.into() {
+                    error!(
+                        self.fsm.peer().logger,
+                        "unexpected term {} more than u32::MAX. Bucket
+                    version will be backward.",
+                        term
+                    );
+                }
+                term << 32
+            };
+            bucket_version
+        };
+
+        let region = self.fsm.peer().region();
+        if util::is_epoch_stale(&region_epoch, region.get_region_epoch()) {
+            error!(
+                 self.fsm.peer().logger,
+                "receive a stale refresh region bucket message";
+                "epoch" => ?region_epoch,
+                "current_epoch" => ?region.get_region_epoch(),
+            );
+
+            return;
+        }
+
+        let mut current_version = self
+            .fsm
+            .peer()
+            .region_buckets()
+            .as_ref()
+            .map(|b| b.meta.version)
+            .unwrap_or_default();
+        if current_version == 0 {
```

Comment:
I don't think this is necessary. If `region_buckets` is `None`, `last_region_buckets` will certainly be `None`.

Reply:
In split region operator , the `region_buckets` should be reset. In this case, the version is saved in `last_region_buckets`.

Reply:
Where exactly? I didn't see mutable accessor for `Option<BucketStat>`

Reply:
[Here it is](https://github.com/tikv/tikv/blob/8d63c2714f0f00fc0a3fe8bb468cbe2aeb851144/components/raftstore-v2/src/operation/command/admin/split.rs#L514 )

Reply:
After this is moved to `Peer`, you can use `self.region_buckets().as_ref().or(self.last_region_buckets().as_ref()`.